modules/repofiles/delete.go

```diff
@@ -0,0 +1,230 @@
+	// If we aren't branching to a new branch, make sure user can commit to the given branch
+	if opts.NewBranch != opts.OldBranch {
+		newBranch, err := repo.GetBranch(opts.NewBranch)
+		if git.IsErrNotExist(err) {
+			return nil, err
+		}
+		if newBranch != nil {
+			return nil, models.ErrBranchAlreadyExists{
+				BranchName: opts.NewBranch,
+			}
+		}
+	} else {
+		if protected, _ := repo.IsProtectedBranchForPush(opts.OldBranch, doer); protected {
+			return nil, models.ErrUserCannotCommit{
+				UserName: doer.LowerName,
+			}
+		}
+	}
+
+	// Check that the path given in opts.treeName is valid (not a git path)
+	treePath := CleanUploadFileName(opts.TreePath)
+	if treePath == "" {
+		return nil, models.ErrFilenameInvalid{
+			Path: opts.TreePath,
+		}
+	}
+
+	message := strings.TrimSpace(opts.Message)
+
+	author, committer := GetAuthorAndCommitterUsers(opts.Committer, opts.Author, doer)
+
+	t, err := NewTemporaryUploadRepository(repo)
+	defer t.Close()
+	if err != nil {
+		return nil, err
+	}
+	if err := t.Clone(opts.OldBranch); err != nil {
+		return nil, err
+	}
+	if err := t.SetDefaultIndex(); err != nil {
+		return nil, err
+	}
+
+	if opts.LastCommitID == "" {
+		commitID, err := t.GetLastCommit()
+		if err != nil {
+			return nil, err
+		}
+		opts.LastCommitID = commitID
+	}
+
+	gitRepo, err := git.OpenRepository(repo.RepoPath())
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the commit of the original branch
+	commit, err := gitRepo.GetBranchCommit(opts.OldBranch)
+	if err != nil {
+		return nil, err // Couldn't get a commit for the branch
+	}
+
+	// Get the files in the index
+	filesInIndex, err := t.LsFiles(opts.TreePath)
+	if err != nil {
+		return nil, fmt.Errorf("DeleteRepoFile: %v", err)
+	}
+
+	// Find the file we want to delete in the index
+	inFilelist := false
+	for _, file := range filesInIndex {
+		if file == opts.TreePath {
+			inFilelist = true
+		}
+	}
+	if !inFilelist {
+		return nil, models.ErrRepoFileDoesNotExist{
+			Path: opts.TreePath,
+		}
+	}
+
+	// Get the entry of treePath and check if the SHA given is the same as the file
+	entry, err := commit.GetTreeEntryByPath(treePath)
+	if err != nil {
+		return nil, err
+	}
+	if opts.SHA != "" {
+		// If a SHA was given and the SHA given doesn't match the SHA of the fromTreePath, throw error
+		if opts.SHA != entry.ID.String() {
+			return nil, models.ErrShaDoesNotMatch{
+				Path:       treePath,
+				GivenSHA:   opts.SHA,
+				CurrentSHA: entry.ID.String(),
+			}
+		}
+	} else if opts.LastCommitID != "" {
+		// If a lastCommitID was given and it doesn't match the commitID of the head of the branch throw
+		// an error, but only if we aren't creating a new branch.
+		if commit.ID.String() != opts.LastCommitID && opts.OldBranch == opts.NewBranch {
+			// CommitIDs don't match, but we don't want to throw a ErrCommitIDDoesNotMatch unless
+			// this specific file has been edited since opts.LastCommitID
+			files, err := commit.GetFilesChangedSinceCommit(opts.LastCommitID)
```

Comment:
This is a bad function to just look for one file diff. The simplest command I can think of is:

`git diff --name-only -z Ref1 Ref2 -- FILE_NAME`

Reply:
Made a new function in the commit class to do this. Please review.

Reply:
https://github.com/go-gitea/gitea/pull/6314/commits/414e6a548483067d69e3a9157cd51986331ab3f0